File: plugins/cordova-plugin-gleap/src/ios/GleapPlugin.m
/********* GleapPlugin.m Cordova Plugin Implementation *******/

#import <Cordova/CDV.h>
#import <Gleap/Gleap.h>

@interface GleapPlugin : CDVPlugin {
  // Member variables go here.
}

- (void)initialize:(CDVInvokedUrlCommand*)command;
- (void)identify:(CDVInvokedUrlCommand*)command;
- (void)startFeedbackFlow:(CDVInvokedUrlCommand*)command;
- (void)startBot:(CDVInvokedUrlCommand*)command;
- (void)sendSilentCrashReport:(CDVInvokedUrlCommand*)command;
- (void)setLanguage:(CDVInvokedUrlCommand*)command;
- (void)open:(CDVInvokedUrlCommand*)command;
- (void)openNews:(CDVInvokedUrlCommand*)command;
- (void)openNewsArticle:(CDVInvokedUrlCommand*)command;
- (void)openHelpCenter:(CDVInvokedUrlCommand*)command;
- (void)openHelpCenterArticle:(CDVInvokedUrlCommand*)command;
- (void)openHelpCenterCollection:(CDVInvokedUrlCommand*)command;
- (void)searchHelpCenter:(CDVInvokedUrlCommand*)command;
- (void)close:(CDVInvokedUrlCommand*)command;
- (void)isOpened:(CDVInvokedUrlCommand*)command;
- (void)trackEvent:(CDVInvokedUrlCommand*)command;
- (void)attachCustomData:(CDVInvokedUrlCommand*)command;
- (void)setCustomData:(CDVInvokedUrlCommand*)command;
- (void)removeCustomData:(CDVInvokedUrlCommand*)command;
- (void)clearCustomData:(CDVInvokedUrlCommand*)command;
- (void)enableDebugConsoleLog:(CDVInvokedUrlCommand*)command;
- (void)clearIdentity:(CDVInvokedUrlCommand*)command;
- (void)showSurvey:(CDVInvokedUrlCommand*)command;
- (void)log:(CDVInvokedUrlCommand*)command;
- (void)preFillForm:(CDVInvokedUrlCommand*)command;
- (void)getIdentity:(CDVInvokedUrlCommand*)command;
- (void)isUserIdentified:(CDVInvokedUrlCommand*)command;
- (void)setDisableInAppNotifications:(CDVInvokedUrlCommand*)command;
@end

@implementation GleapPlugin

- (void)initialize:(CDVInvokedUrlCommand*)command
{
    @try {
        CDVPluginResult* pluginResult = nil;
        
        if (command.arguments.count == 0) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        } else {
            NSString* token = [command.arguments objectAtIndex: 0];
            [Gleap initializeWithToken: token];
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        }
        
        [Gleap setApplicationType: CORDOVA];

        [self.commandDelegate sendPluginResult: pluginResult callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)showSurvey:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* surveyId = [command.arguments objectAtIndex: 0];
        NSString* format = [command.arguments objectAtIndex: 1];

        GleapSurveyFormat surveyFormat = SURVEY;
        if (format != nil && [format isEqualToString: @"survey_full"]) {
            surveyFormat = SURVEY_FULL;
        }
        
        [Gleap showSurvey: surveyId andFormat: surveyFormat];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)clearIdentity:(CDVInvokedUrlCommand *)command {
    @try {
        [Gleap clearIdentity];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)setTags:(CDVInvokedUrlCommand *)command {
    @try {
        NSArray* tags = [command.arguments objectAtIndex: 0];

        [Gleap setTags: tags];

        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)isUserIdentified:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString: [Gleap isUserIdentified] ? @"true" : @"false"] callbackId:command.callbackId];
}

- (void)getIdentity:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: [Gleap getIdentity]] callbackId:command.callbackId];
}

- (void)log:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* message = [command.arguments objectAtIndex: 0];
        NSString* logLevel = @"";
        
        if (command.arguments.count > 1) {
            logLevel = [command.arguments objectAtIndex: 1];
        }
        
        GleapLogLevel logLevelObj = INFO;
        if ([logLevel isEqualToString: @"ERROR"]) {
            logLevelObj = ERROR;
        }
        if ([logLevel isEqualToString: @"WARNING"]) {
            logLevelObj = WARNING;
        }
        
        [Gleap log: message withLogLevel: logLevelObj];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)preFillForm:(CDVInvokedUrlCommand *)command {
    @try {
        NSDictionary* data = [command.arguments objectAtIndex: 0];
        
        [Gleap preFillForm: data];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)identify:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* userId = [command.arguments objectAtIndex: 0];
        NSDictionary* userData = [command.arguments objectAtIndex: 1];
        NSString* userHash = [command.arguments objectAtIndex: 2];
        
        GleapUserProperty *gleapUserData = [[GleapUserProperty alloc] init];
        if (userData != nil) {
            if ([userData objectForKey: @"email"]) {
                gleapUserData.email = [userData objectForKey: @"email"];
            }
            if ([userData objectForKey: @"phone"]) {
                gleapUserData.phone = [userData objectForKey: @"phone"];
            }
            if ([userData objectForKey: @"value"]) {
                gleapUserData.value = [userData objectForKey: @"value"];
            }
            if ([userData objectForKey: @"name"]) {
                gleapUserData.name = [userData objectForKey: @"name"];
            }
            if ([userData objectForKey: @"customData"]) {
                gleapUserData.customData = [userData objectForKey: @"customData"];
            }
        }
        
        if (userHash == nil) {
            [Gleap identifyUserWith: userId andData: gleapUserData];
        } else {
            [Gleap identifyUserWith: userId andData: gleapUserData andUserHash: userHash];
        }

        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)sendSilentCrashReport:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* description = [command.arguments objectAtIndex: 0];
        NSString* severity = [command.arguments objectAtIndex: 1];
        NSDictionary* excludeData = [command.arguments objectAtIndex: 2];
        
        GleapBugSeverity prio = MEDIUM;
        if ([severity isEqualToString: @"LOW"]) {
            prio = LOW;
        }
        if ([severity isEqualToString: @"HIGH"]) {
            prio = HIGH;
        }
        
        if (excludeData == nil || [excludeData isEqual: [NSNull null]]) {
            excludeData = [[NSDictionary alloc] init];
        }
        
        [Gleap sendSilentCrashReportWith: description andSeverity: prio andDataExclusion: excludeData andCompletion:^(bool success) {
            if (success) {
                [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
            } else {
                [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR] callbackId:command.callbackId];
            }
        }];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)setLanguage:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* language = [command.arguments objectAtIndex: 0];
        if (language != nil) {
            [Gleap setLanguage: language];
        }

        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)open:(CDVInvokedUrlCommand *)command {
    [Gleap open];
    
    [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
}

- (void)setDisableInAppNotifications:(CDVInvokedUrlCommand *)command {
    @try {
        bool disableInAppNotifications = [[command.arguments objectAtIndex: 0] boolValue];

        [Gleap setDisableInAppNotifications: disableInAppNotifications];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)openFeatureRequests:(CDVInvokedUrlCommand *)command {
    @try {
        bool showBackButton = [[command.arguments objectAtIndex: 0] boolValue];

        [Gleap openFeatureRequests: showBackButton];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)openNews:(CDVInvokedUrlCommand *)command {
    @try {
        bool showBackButton = [[command.arguments objectAtIndex: 0] boolValue];

        [Gleap openNews: showBackButton];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)openNewsArticle:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* articleId = [command.arguments objectAtIndex: 0];
        bool showBackButton = [[command.arguments objectAtIndex: 1] boolValue];

        [Gleap openNewsArticle: articleId andShowBackButton: showBackButton];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)openHelpCenter:(CDVInvokedUrlCommand *)command {
    @try {
        bool showBackButton = [[command.arguments objectAtIndex: 0] boolValue];

        [Gleap openHelpCenter: showBackButton];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)openHelpCenterArticle:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* articleId = [command.arguments objectAtIndex: 0];
        bool showBackButton = [[command.arguments objectAtIndex: 1] boolValue];

        [Gleap openHelpCenterArticle: articleId andShowBackButton: showBackButton];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)openHelpCenterCollection:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* collectionId = [command.arguments objectAtIndex: 0];
        bool showBackButton = [[command.arguments objectAtIndex: 1] boolValue];

        [Gleap openHelpCenterCollection: collectionId andShowBackButton: showBackButton];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)searchHelpCenter:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* term = [command.arguments objectAtIndex: 0];
        bool showBackButton = [[command.arguments objectAtIndex: 1] boolValue];

        [Gleap searchHelpCenter: term andShowBackButton: showBackButton];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)close:(CDVInvokedUrlCommand *)command {
    [Gleap close];
    
    [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
}

- (void)isOpened:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus: CDVCommandStatus_OK messageAsString: [Gleap isOpened] ? @"true" : @"false"] callbackId:command.callbackId];
}

- (void)trackEvent:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* name = [command.arguments objectAtIndex: 0];
        NSDictionary* data = [command.arguments objectAtIndex: 1];
        if (name != nil) {
            if (data != nil) {
                [Gleap trackEvent: name withData: data];
            } else {
                [Gleap trackEvent: name];
            }
        }

        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)startBot:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* botId = [command.arguments objectAtIndex: 0];
        bool showBackButton = [[command.arguments objectAtIndex: 1] boolValue];
        if (botId != nil) {
            [Gleap startBot: botId showBackButton: showBackButton];
        }

        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)startFeedbackFlow:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* feedbackFlow = [command.arguments objectAtIndex: 0];
        bool showBackButton = [[command.arguments objectAtIndex: 1] boolValue];
        if (feedbackFlow != nil) {
            [Gleap startFeedbackFlow: feedbackFlow showBackButton: showBackButton];
        }

        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)showFeedbackButton:(CDVInvokedUrlCommand *)command {
    @try {
        bool showBackButton = [[command.arguments objectAtIndex: 0] boolValue];
        
        [Gleap showFeedbackButton: showBackButton];
        
        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)attachCustomData:(CDVInvokedUrlCommand *)command {
    @try {
        NSDictionary* customData = [command.arguments objectAtIndex: 0];
        if (customData != nil) {
            [Gleap attachCustomData: customData];
        }

        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)setCustomData:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* key = [command.arguments objectAtIndex: 0];
        NSString* value = [command.arguments objectAtIndex: 1];
        if (key != nil && value != nil) {
            [Gleap setCustomData: value forKey: key];
        }

        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)removeCustomData:(CDVInvokedUrlCommand *)command {
    @try {
        NSString* key = [command.arguments objectAtIndex: 0];
        if (key != nil) {
            [Gleap removeCustomDataForKey: key];
        }

        [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
    }
    @catch (NSException *exception) {
        NSLog(@"GleapPlugin: Wrong arguments passed.");
    }
}

- (void)clearCustomData:(CDVInvokedUrlCommand *)command {
    [Gleap clearCustomData];

    [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
}

- (void)enableDebugConsoleLog:(CDVInvokedUrlCommand *)command {
    [Gleap enableDebugConsoleLog];

    [self.commandDelegate sendPluginResult: [CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
}

@end

File: plugins/cordova-plugin-gleap/src/android/GleapPlugin.java
package io.cordova.gleap;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import io.gleap.Gleap;
import io.gleap.GleapUser;
import io.gleap.SurveyType;
import io.gleap.GleapUserProperties;
import io.gleap.APPLICATIONTYPE;
import io.gleap.PrefillHelper;
import io.gleap.GleapLogLevel;

/**
 * This class echoes a string called from JavaScript.
 */
public class GleapPlugin extends CordovaPlugin {

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("initialize")) {
            String message = args.getString(0);
            this.initialize(message, callbackContext);
            return true;
        }
        if (action.equals("identify")) {
            this.identify(args);
            return true;
        }
        if (action.equals("setLanguage")) {
            String language = args.getString(0);
            this.setLanguage(language);
            return true;
        }
        if (action.equals("open")) {
            this.open();
            return true;
        }
        if (action.equals("showFeedbackButton")) {
            this.showFeedbackButton(args);
            return true;
        }
        if (action.equals("setDisableInAppNotifications")) {
            this.setDisableInAppNotifications(args);
            return true;
        }
        if (action.equals("openNews")) {
            this.openNews(args);
            return true;
        }
        if (action.equals("openFeatureRequests")) {
            this.openFeatureRequests(args);
            return true;
        }
        if (action.equals("openNewsArticle")) {
            this.openNewsArticle(args);
            return true;
        }
        if (action.equals("openHelpCenter")) {
            this.openHelpCenter(args);
            return true;
        }
        if (action.equals("openHelpCenterArticle")) {
            this.openHelpCenterArticle(args);
            return true;
        }
        if (action.equals("openHelpCenterCollection")) {
            this.openHelpCenterCollection(args);
            return true;
        }
        if (action.equals("searchHelpCenter")) {
            this.searchHelpCenter(args);
            return true;
        }
        if (action.equals("sendSilentCrashReport")) {
            this.sendSilentCrashReport(args);
            return true;
        }
        if (action.equals("close")) {
            this.close();
            return true;
        }
        if (action.equals("isOpened")) {
            return this.isOpened(callbackContext);
        }
        if (action.equals("startFeedbackFlow")) {
            this.startFeedbackFlow(args);
            return true;
        }
        if (action.equals("startBot")) {
            this.startBot(args);
            return true;
        }
        if (action.equals("trackEvent")) {
            this.trackEvent(args);
            return true;
        }
        if (action.equals("attachCustomData")) {
            this.attachCustomData(args);
            return true;
        }
        if (action.equals("setCustomData")) {
            this.setCustomData(args);
            return true;
        }
        if (action.equals("clearIdentity")) {
            this.clearIdentity();
            return true;
        }
        if (action.equals("removeCustomData")) {
            this.removeCustomData(args);
            return true;
        }
        if (action.equals("showSurvey")) {
            this.showSurvey(args);
            return true;
        }
        if (action.equals("log")) {
            this.log(args);
            return true;
        }
        if (action.equals("preFillForm")) {
            this.preFillForm(args);
            return true;
        }
        if (action.equals("clearCustomData")) {
            this.clearCustomData();
            return true;
        }
        if (action.equals("enableDebugConsoleLog")) {
            this.enableDebugConsoleLog();
            return true;
        }
        if (action.equals("getIdentity")) {
            this.getIdentity(callbackContext);
            return true;
        }
        if (action.equals("isUserIdentified")) {
            this.isUserIdentified(callbackContext);
            return true;
        }
        if (action.equals("setTags")) {
            this.setTags(args);
            return true;
        }
        return false;
    }

    private void initialize(String token, CallbackContext callbackContext) {
        if (token != null && token.length() > 0) {
            Gleap.initialize(token, this.cordova.getActivity().getApplication());
            Gleap.getInstance().setApplicationType(APPLICATIONTYPE.CORDOVA);
            callbackContext.success(token);
        } else {
            callbackContext.error("Expected one non-empty string argument.");
        }
    }

    private void setTags(JSONArray args) {
        try {
            JSONArray jsonTags = args.getJSONArray(0);
            if (jsonTags != null) {
                String[] tags = new String[jsonTags.length()];
                for (int i = 0; i < jsonTags.length(); i++) {
                    tags[i] = jsonTags.getString(i);
                }
                Gleap.getInstance().setTags(tags);
            }
        } catch (Exception ex) {
        }
    }

    private void preFillForm(JSONArray args) {
        try {
            JSONObject prefillData = args.getJSONObject(0);

            PrefillHelper.getInstancen().setPrefillData(prefillData);
        } catch (Exception ex) {
        }
    }

    private void identify(JSONArray args) {
        try {
            String userId = args.getString(0);
            JSONObject userData = args.getJSONObject(1);
            String userHash = args.getString(2);

            GleapUserProperties gleapUserProperties = new GleapUserProperties(userId);

            if (userData != null) {
                if (userData.has("email")) {
                    gleapUserProperties.setEmail(userData.getString("email"));
                }
                if (userData.has("phone")) {
                    gleapUserProperties.setPhoneNumber(userData.getString("phone"));
                }
                if (userData.has("value")) {
                    gleapUserProperties.setValue(Double.parseDouble(userData.getString("value")));
                }
                if (userData.has("name")) {
                    gleapUserProperties.setName(userData.getString("name"));
                }
                if (userData.has("customData")) {
                    gleapUserProperties.setCustomData(userData.getJSONObject("customData"));
                }
            }

            Gleap.getInstance().identifyUser("13", gleapUserProperties);
        } catch (Exception ex) {
        }
    }

    private void log(JSONArray args) {
        try {
            String message = args.getString(0);
            GleapLogLevel logLevelObj = GleapLogLevel.INFO;

            try {
                String logLevel = args.getString(1);
                if (logLevel != null && logLevel.equals("ERROR")) {
                    logLevelObj = GleapLogLevel.ERROR;
                }
                if (logLevel != null && logLevel.equals("WARNING")) {
                    logLevelObj = GleapLogLevel.WARNING;
                }
            } catch (Exception ex) {
            }

            Gleap.getInstance().log(message, logLevelObj);
        } catch (Exception ex) {
        }
    }

    private void showSurvey(JSONArray args) {
        try {
            String surveyId = args.getString(0);
            String format = args.getString(1);
            SurveyType surveyFormat;
            switch (format) {
                case "survey_full":
                surveyFormat = SurveyType.SURVEY_FULL;
                break;
                default:
                surveyFormat = SurveyType.SURVEY;
            }

            Gleap.getInstance().showSurvey(surveyId, surveyFormat);
        } catch (Exception ex) {
        }
    }

    private void setLanguage(String language) {
        Gleap.getInstance().setLanguage(language);
    }

    private void open() {
        Gleap.getInstance().open();
    }

    private void openFeatureRequests(JSONArray args) {
        try {
            Boolean showBackButton = args.getBoolean(0);

            Gleap.getInstance().openFeatureRequests(showBackButton);
        } catch (Exception ex) {
        }
    }

    private void setDisableInAppNotifications(JSONArray args) {
        try {
            Boolean disableInAppNotifications = args.getBoolean(0);

            Gleap.getInstance().setDisableInAppNotifications(disableInAppNotifications);
        } catch (Exception ex) {
        }
    }

    private void openNews(JSONArray args) {
        try {
            Boolean showBackButton = args.getBoolean(0);

            Gleap.getInstance().openNews(showBackButton);
        } catch (Exception ex) {
        }
    }

    private void openNewsArticle(JSONArray args) {
        try {
            String articleId = args.getString(0);
            Boolean showBackButton = args.getBoolean(1);

            Gleap.getInstance().openNewsArticle(articleId, showBackButton);
        } catch (Exception ex) {
        }
    }

    private void openHelpCenter(JSONArray args) {
        try {
            Boolean showBackButton = args.getBoolean(0);

            Gleap.getInstance().openHelpCenter(showBackButton);
        } catch (Exception ex) {
        }
    }

    private void openHelpCenterArticle(JSONArray args) {
        try {
            String articleId = args.getString(0);
            Boolean showBackButton = args.getBoolean(1);

            Gleap.getInstance().openHelpCenterArticle(articleId, showBackButton);
        } catch (Exception ex) {
        }
    }

    private void openHelpCenterCollection(JSONArray args) {
        try {
            String collectionId = args.getString(0);
            Boolean showBackButton = args.getBoolean(1);

            Gleap.getInstance().openHelpCenterCollection(collectionId, showBackButton);
        } catch (Exception ex) {
        }
    }

    private void searchHelpCenter(JSONArray args) {
        try {
            String term = args.getString(0);
            Boolean showBackButton = args.getBoolean(1);

            Gleap.getInstance().searchHelpCenter(term, showBackButton);
        } catch (Exception ex) {
        }
    }

    private void clearIdentity() {
        Gleap.getInstance().clearIdentity();
    }

    private void getIdentity(CallbackContext callbackContext) {
        JSONObject identityObj = new JSONObject();

        try {
            GleapUser gleapUser = Gleap.getInstance().getIdentity();
            if (gleapUser != null) {
                GleapUserProperties userProps = gleapUser.getGleapUserProperties();

                if (userProps != null) {
                    identityObj.put("userId", gleapUser.getUserId());
                    identityObj.put("phone", userProps.getPhoneNumber());
                    identityObj.put("email", userProps.getEmail());
                    identityObj.put("name", userProps.getName());
                    identityObj.put("value", userProps.getValue());
                }
            }
        } catch (Exception ex) {}

        callbackContext.success(identityObj);
    }

    private void isUserIdentified(CallbackContext callbackContext) {
        callbackContext.success(Gleap.getInstance().isUserIdentified() ? 1 : 0);
    }

    private void sendSilentCrashReport(JSONArray args) {
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {

                try {
                    String description = args.getString(0);
                    Gleap.SEVERITY severity = Gleap.SEVERITY.MEDIUM;
                    if (args.getString(1) == "LOW") {
                        severity = Gleap.SEVERITY.LOW;
                    }
                    if (args.getString(1) == "HIGH") {
                        severity = Gleap.SEVERITY.HIGH;
                    }

                    JSONObject exclude = args.getJSONObject(2);
                    if (exclude == null) {
                        exclude = new JSONObject();
                    }

                    Gleap.getInstance().sendSilentCrashReport(description, severity,
                            exclude);
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        });
    }

    private void close() {
        Gleap.getInstance().close();
    }

    private boolean isOpened(CallbackContext callbackContext) {
        callbackContext.success(Gleap.getInstance().isOpened() ? 1 : 0);
        return true;
    }

    private void showFeedbackButton(JSONArray args) {
        try {
            Boolean show = args.getBoolean(0);
            Gleap.getInstance().showFeedbackButton(show);
        } catch (Exception ex) {
        }
    }

    private void startBot(JSONArray args) {
        try {
            String botId = args.getString(0);
            Boolean showBackButton = args.getBoolean(1);
            Gleap.getInstance().startBot(botId, showBackButton);
        } catch (Exception ex) {
        }
    }
    private void startFeedbackFlow(JSONArray args) {
        try {
            String feedbackFlow = args.getString(0);
            Boolean showBackButton = args.getBoolean(1);
            Gleap.getInstance().startFeedbackFlow(feedbackFlow, showBackButton);
        } catch (Exception ex) {
        }
    }

    private void trackEvent(JSONArray args) {
        try {
            String name = args.getString(0);
            JSONObject data = args.getJSONObject(1);

            if (data != null) {
                Gleap.getInstance().trackEvent(name, data);
            } else {
                Gleap.getInstance().trackEvent(name);
            }
        } catch (Exception ex) {
            System.out.println(ex);
        }
    }

    private void attachCustomData(JSONArray args) {
        try {
            JSONObject data = args.getJSONObject(0);
            Gleap.getInstance().attachCustomData(data);
        } catch (Exception ex) {
            System.out.println(ex);
        }
    }

    private void setCustomData(JSONArray args) {
        try {
            String key = args.getString(0);
            String value = args.getString(1);
            Gleap.getInstance().setCustomData(key, value);
        } catch (Exception ex) {
            System.out.println(ex);
        }
    }

    private void removeCustomData(JSONArray args) {
        try {
            String key = args.getString(0);
            Gleap.getInstance().removeCustomDataForKey(key);
        } catch (Exception ex) {
        }
    }

    private void clearCustomData() {
        Gleap.getInstance().clearCustomData();
    }

    private void enableDebugConsoleLog() {
    }
}
